src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1330,33 +1342,37 @@ static ucs_status_t
 uct_ib_mlx5_devx_device_mem_free(uct_md_h uct_md, uct_mem_h memh)
 {
 #if HAVE_IBV_DM
-    uct_ib_mlx5_devx_mem_t *ib_memh = memh;
-    struct ibv_mr *ib_mr            = ib_memh->mrs[UCT_IB_MR_DEFAULT].super.ib;
-    size_t length                   = ib_mr->length;
+    uct_ib_mlx5_devx_mem_t *ib_memh        = memh;
+    size_t length                          = ib_mr->length;
+    struct ibv_dm *dm                      = ib_memh->dm;
+    uct_md_mem_dereg_params_t dereg_params = {
+        .field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH |
+                      UCT_MD_MEM_DEREG_FIELD_FLAGS,
+        .memh       = memh,
+        .flags      = 0
+    };
     ucs_status_t status;
     int ret;
 
     uct_ib_mlx5_devx_obj_destroy(ib_memh->dm_addr_dvmr, "DM-KSM");
 
-    status = uct_ib_dereg_mr(ib_mr);
-    if (status != UCS_OK) {
-        ucs_warn("%s: failed to dereg device memory mr",
-                 ucs_status_string(status));
-    }
-
     ret = munmap(ib_memh->address, length);
     if (ret != 0) {
         ucs_warn("munmap(address=%p, length=%zu) failed: %m", ib_memh->address,
                  length);
     }
 
-    ret = UCS_PROFILE_CALL(ibv_free_dm, ib_memh->dm);
+    status = uct_ib_mlx5_devx_mem_dereg(uct_md, &dereg_params);
+    if (status != UCS_OK) {
+        ucs_warn("%s: uct_ib_mlx5_devx_mem_dereg() failed", ucs_status_string(status));
+    }
```

Comment:
though it works, I think it would be nicer to have a common helper function for both uct_ib_mlx5_devx_device_mem_free and uct_ib_mlx5_devx_mem_dereg. This common function will release only what is really relevant for both cases. For example, it will not check UCT_IB_MEM_IMPORTED, memh->umem